server/core/broker_key_generator.go

```diff
@@ -0,0 +1,197 @@
+//Copyright 2017 Huawei Technologies Co., Ltd
+//
+//Licensed under the Apache License, Version 2.0 (the "License");
+//you may not use this file except in compliance with the License.
+//You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+//Unless required by applicable law or agreed to in writing, software
+//distributed under the License is distributed on an "AS IS" BASIS,
+//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//See the License for the specific language governing permissions and
+//limitations under the License.
+package core
+
+import (
+	"strconv"
+
+	"github.com/ServiceComb/service-center/pkg/util"
+)
+
+const (
+	BROKER_ROOT_KEY              = "cse-pact"
+	BROKER_PARTICIPANT_KEY       = "participant"
+	BROKER_VERSION_KEY           = "version"
+	BROKER_PACT_KEY              = "pact"
+	BROKER_PACT_VERSION_KEY      = "pact-version"
+	BROKER_PACT_TAG_KEY          = "pact-tag"
+	BROKER_PACT_VERIFICATION_KEY = "verification"
+	BROKER_PACT_LATEST           = "latest"
+)
+
+// GetBrokerRootKey returns url (/cse-pact)
+func GetBrokerRootKey() string {
+	return util.StringJoin([]string{
+		"",
+		BROKER_ROOT_KEY,
+	}, "/")
+}
+
+//GetBrokerLatestKey returns  pact related keys
+func GetBrokerLatestKey(tenant string) string {
+	return util.StringJoin([]string{
+		GetBrokerRootKey(),
+		BROKER_PACT_LATEST,
+		tenant,
+	}, "/")
+}
+
+//GetBrokerParticipantKey returns the participant root key
+func GetBrokerParticipantKey(tenant string) string {
+	return util.StringJoin([]string{
+		GetBrokerRootKey(),
+		BROKER_PARTICIPANT_KEY,
+		tenant,
+	}, "/")
+}
+
+//GenerateBrokerParticipantKey returns the participant key
+func GenerateBrokerParticipantKey(tenant string, appId string, serviceName string) string {
+	return util.StringJoin([]string{
+		GetBrokerParticipantKey(tenant),
+		appId,
+		serviceName,
+	}, "/")
+}
+
+//GetBrokerVersionKey returns th root version key
+func GetBrokerVersionKey(tenant string) string {
+	return util.StringJoin([]string{
+		GetBrokerRootKey(),
+		BROKER_VERSION_KEY,
+		tenant,
+	}, "/")
+}
+
+//GenerateBrokerVersionKey returns the version key
+func GenerateBrokerVersionKey(tenant string, number string, participantId int32) string {
+	return util.StringJoin([]string{
+		GetBrokerVersionKey(tenant),
+		number,
+		strconv.Itoa(int(participantId)),
+	}, "/")
+}
+
+//GetBrokerPactKey returns the pact root key
+func GetBrokerPactKey(tenant string) string {
+	return util.StringJoin([]string{
+		GetBrokerRootKey(),
+		BROKER_PACT_KEY,
+		tenant,
+	}, "/")
+}
+
+//GenerateBrokerPactKey returns the pact key
+func GenerateBrokerPactKey(tenant string, consumerParticipantId int32,
+	providerParticipantId int32, sha []byte) string {
+	return util.StringJoin([]string{
+		GetBrokerPactKey(tenant),
+		strconv.Itoa(int(consumerParticipantId)),
+		strconv.Itoa(int(providerParticipantId)),
+		string(sha),
```

Comment:
'sha' may be encoded in unicode that lead to get a wrong result from etcd server.